### PR TITLE
Making driver default compliant to JDBC 4.2 Specs

### PR DESCRIPTION
**Pull request by cheenamalhotra**
Initial changes for JDBC 4.2 compliance - making default compliance level after removing JDK 7 support.

### CODE REVIEW

File: src/main/java/com/microsoft/sqlserver/jdbc/ISQLServerCallableStatement.java [hunk 1]
```diff
@@ -8,49 +8,1086 @@
...
      */
-    public microsoft.sql.DateTimeOffset getDateTimeOffset(int parameterIndex) throws SQLException;
+    public microsoft.sql.DateTimeOffset getDateTimeOffset(int parameterIndex) throws SQLServerException;
 
     /**
      * Gets the DateTimeOffset value of parameter with name parameterName
      * 
      * @param parameterName
      *            the name of the parameter
      * @return DateTimeOffset value
```
Comment (rene-ye) [inline src/main/java/com/microsoft/sqlserver/jdbc/ISQLServerCallableStatement.java]: most other return javadocs have a small description of the value, should we add one to this as well to maintain consistency?

File: src/test/java/com/microsoft/sqlserver/jdbc/bulkCopy/ISQLServerBulkRecordIssuesTest.java [hunk 2]
```diff
@@ -445,4 +408,54 @@ public boolean next() throws SQLServerException {
...
     }
 
+    @Override
+    public void addColumnMetadata(int positionInFile,
+            String name,
+            int jdbcType,
+            int precision,
+            int scale,
+            DateTimeFormatter dateTimeFormatter) throws SQLServerException {
+        // TODO Not Implemented
```
Comment (rene-ye) [inline src/test/java/com/microsoft/sqlserver/jdbc/bulkCopy/ISQLServerBulkRecordIssuesTest.java]: Should we throw a "not implemented" exception when users try to use these APIs; like we do with sharding APIs.

File: src/main/java/com/microsoft/sqlserver/jdbc/Geography.java [hunk 3]
```diff
@@ -15,7 +15,13 @@
...
      * Private constructor used for creating a Geography object from WKT and srid.
+     * 
+     * @param WellKnownText
+     *              Well-Known Text (WKT) provided by the user.
+     * @param srid 
+     *             Spatial Reference Identifier (SRID) provided by the user.
      * @throws SQLServerException 
+     *              if an exception occurs
      */
     private Geography(String WellKnownText, int srid) throws SQLServerException {
```
Comment (lilgreenbird) [inline src/main/java/com/microsoft/sqlserver/jdbc/Geography.java]: seems a bit inconsistent - why is this spelled out "WellKnownText" when the others are all wkt?

File: src/main/java/com/microsoft/sqlserver/jdbc/ISQLServerDataSource.java [hunk 4]
```diff
@@ -112,11 +115,36 @@
...
+     * Retrieves the TransparentNetworkIPResolution value.
+     * 
+     * @return if enabled, returns true. Otherwise, false.
+     */
+    public boolean getTransparentNetworkIPResolution();
+
     /**
      * Sets a Boolean value that indicates if the trustServerCertificate property is enabled.
      * 
      * @param e
```
Comment (lilgreenbird) [inline src/main/java/com/microsoft/sqlserver/jdbc/ISQLServerDataSource.java]: better variable name than "e"?

File: src/main/java/com/microsoft/sqlserver/jdbc/ISQLServerDataSource.java [hunk 5]
```diff
@@ -410,6 +468,21 @@
      */
     public void setAuthenticationScheme(String authenticationScheme);
 
+    /**
+     * sets the authentication mode
+     * 
+     * @param authentication
+     *            the authentication mode
+     */
+    public void setAuthentication(String authentication);
```
Comment (lilgreenbird) [inline src/main/java/com/microsoft/sqlserver/jdbc/ISQLServerDataSource.java]: similar to above, why isn't this setAuthenticationMode(String authenticationMode)?

File: src/main/java/com/microsoft/sqlserver/jdbc/ISQLServerDataSource.java [hunk 6]
```diff
@@ -424,4 +497,297 @@
...
+     * 
+     * @return GSSCredential
+     */
+    public GSSCredential getGSSCredentials();
+
+    /**
+     * Sets the access token.
+     * 
+     * @param accessToken
+     *            to be set in the string property.
```
Comment (lilgreenbird) [inline src/main/java/com/microsoft/sqlserver/jdbc/ISQLServerDataSource.java]: A string that contains the access token?

File: src/main/java/com/microsoft/sqlserver/jdbc/ISQLServerDataSource.java [hunk 7]
```diff
@@ -424,4 +497,297 @@
...
+
+    /**
+     * Getting the cancel timeout
+     * 
+     * @return the number of seconds to wait before we wait for the query timeout to happen.
+     */
+    public int getCancelQueryTimeout();
+
+    /**
+     * If this configuration is false the first execution of a prepared statement will call sp_executesql and not prepare a statement, once the second
```
Comment (lilgreenbird) [inline src/main/java/com/microsoft/sqlserver/jdbc/ISQLServerDataSource.java]: start with description as in other methods?
ie This setting controls blah blah...

File: src/main/java/com/microsoft/sqlserver/jdbc/ISQLServerDataSource.java [hunk 8]
```diff
@@ -424,4 +497,297 @@
...
+     * execution happens it will call sp_prepexec and actually setup a prepared statement handle. Following executions will call sp_execute. This
+     * relieves the need for sp_unprepare on prepared statement close if the statement is only executed once.
+     * 
+     * @param enablePrepareOnFirstPreparedStatementCall
+     *            Changes the setting per the description.
+     */
+    public void setEnablePrepareOnFirstPreparedStatementCall(boolean enablePrepareOnFirstPreparedStatementCall);
+
+    /**
+     * If this configuration returns false the first execution of a prepared statement will call sp_executesql and not prepare a statement, once the
```
Comment (lilgreenbird) [inline src/main/java/com/microsoft/sqlserver/jdbc/ISQLServerDataSource.java]: Retrieves the setting which controls blah...

File: src/main/java/com/microsoft/sqlserver/jdbc/ISQLServerDataSource.java [hunk 9]
```diff
@@ -424,4 +497,297 @@
...
+     * prepared statement close. If it is set to {@literal >} 1 these calls will be batched together to avoid overhead of calling sp_unprepare too
+     * often.
+     * 
+     * @param serverPreparedStatementDiscardThreshold
+     *            Changes the setting per the description.
+     */
+    public void setServerPreparedStatementDiscardThreshold(int serverPreparedStatementDiscardThreshold);
+
+    /**
+     * This setting controls how many outstanding prepared statement discard actions (sp_unprepare) can be outstanding per connection before a call to
```
Comment (lilgreenbird) [inline src/main/java/com/microsoft/sqlserver/jdbc/ISQLServerDataSource.java]: Retrieves the setting which controls blah...

File: src/main/java/com/microsoft/sqlserver/jdbc/ISQLServerPreparedStatement.java [hunk 10]
```diff
@@ -18,10 +24,819 @@
...
+     *            the parameter value
+     * @param forceEncrypt
+     *            If the boolean forceEncrypt is set to true, the query parameter will only be set if the designation column is encrypted and Always
+     *            Encrypted is enabled on the connection or on the statement. If the boolean forceEncrypt is set to false, the driver will not force
+     *            encryption on parameters.
+     * @throws SQLServerException
+     *             when an error occurs
+     */
+    public void setNString(int parameterIndex,
+            String value,
```
Comment (lilgreenbird) [inline src/main/java/com/microsoft/sqlserver/jdbc/ISQLServerPreparedStatement.java]: use str so it's similar to setString?

**Pull request merged**